feat(widget): add support for QtKeychain to store/retrieve passwords

thfi opened the pull request:
This commit adds support for storing to and retrieving passwords from a operating system or desktop environment-specific secure storage. It makes use of [Frank Osterfeld's QtKeychain](https://github.com/frankosterfeld/qtkeychain), which make store passwords in Mac OS X's Keychain, GNOME's keyring, KDE's wallet, or Windows' Credential Store.
Code is tested on Linux using KDE's wallet. The code is optional, to activate it, `QTOX_QTKEYCHAIN` must be defined when calling qmake.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tux3/qtox/3332)

<!-- Reviewable:end -->

zetok commented:
- [x] missing docs https://github.com/tux3/qTox/blob/75c72a4ebd1fdc75ff3bbf8cbcac49ef8df39fcf/INSTALL.md#dependencies
- [x] even if support for it is optional, it should be build with and without it on travis, currently it's only built without

sudden6 commented:
Please change your commit messages according to https://github.com/tux3/qTox/blob/master/CONTRIBUTING.md#commit-message-format then the travis build should work :)

agilob commented:
I like this change, but I think it should introduce new setting, I don't want to store some credentials in KWallet, as I understand this feature forces usage of system-keychains.

zetok commented:
@thfi 

> (…) as I understand this feature forces usage of system-keychains.

Needs to be changed – i.e. even when compiled with support for it, user shouldn't be forced to use it. And the default setting should be to not use it, since users generally don't know how to use that stuff.

Also, PR needs a rebase.

thfi committed:
feat(widget): add support for QtKeychain to store/retrieve passwords

This commit adds support for QtKeychain, a platform-independent Qt API
for storing passwords securely. According to its documentation, it
stores passwords depending on operating system and desktop environment:
- Mac OS X's Keychain
- Windows' Credential Store
- Linux/KDE: KWallet
- Linux/Gnome: Keyring

Using this library is optional and can be enabled/disabled at compile
time: The feature is enabled by default, but can be disabled by
passing 'DISABLE_QTKEYCHAIN=YES' to a qmake call.

Even if qTox is compiled against QtKeychain, the user can enable/
disable storing the password in the login screen (global setting, not
per profile).
Later, in the privacy settings, the user can choose to clear the
current profile's stored password.

thfi commented:
> Needs to be changed – i.e. even when compiled with support for it, user shouldn't be forced to use it. And the default setting should be to not use it, since users generally don't know how to use that stuff.

I see your point, will work on that. Right now, I am rebasing my patch to the latest code from master.

zetok commented:
Hi,

As you might be aware, repo has been moved under `qTox` org on github.

What does it have to do with your PR, you might ask.

Well, it doesn't change all that much, except that workflow has been slightly improved – now there are required checks that PR needs to pass in order to be merged.

Those checks are:
- travis (checks whether PR builds & commit message format)
- reviewable (requirement for PR to be reviewed & accepted)

Since osx travis was broken for a bit, you will need to rebase your PR on newest master to get it to pass travis check. Sorry for inconvenience.

Reminder: as always, PR needs to be possible to merge without conflicts. Make sure it does ;)

thfi commented:
I made some changes to the code. Most changes concern the user interface, where QtKeychain support can be toggled (login screen, global setting) and the stored password cleared (privacy settings, per profile).

zetok commented:
I didn't review the code, since I don't know C++.

---

Reviewed 2 of 17 files at r1.
Review status: 2 of 17 files reviewed at latest revision, 10 unresolved discussions.

---

_[qtox.pro, line 71 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPQzJHWEiGpMVXPCIf-:-KPQzJHWEiGpMVXPCIf0:b-8sjrm1) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/qtox.pro#L71)):_

> ``` QMake
> }
> 
> contains(DISABLE_QTKEYCHAIN, YES) {
> ```

can be simplified to

```
!contains(DISABLE_QTKEYCHAIN, YES) {
    DEFINES += QTOX_QTKEYCHAIN
}
```

---

_[qtox.pro, line 84 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgWzr0XXok_mzpGK5j:-KPgWzr13waO1TTJw3kU:b-lavkcg) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/qtox.pro#L84)):_

> ``` QMake
>         # Locating QtKeychain on Mac OS X installed with brew
>         LIBS += -L/usr/local/Cellar/qtkeychain/0.6.2/lib/
>         INCLUDEPATH += /usr/local/Cellar/qtkeychain/0.6.2/include/
> ```

This is going to horribly break whenever brew decides to bump qtkeychain version. Is it possible to not hardcode version?

---

_[.travis/build-ubuntu-14-04.sh, line 58 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgX_qpxzin5FVd61Fa:-KPgX_qpxzin5FVd61Fb:b478k8q) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/.travis/build-ubuntu-14-04.sh#L58)):_

> ``` Shell
> tar xf QtKeychain-${QTKEYCHAIN_VERSION}.tar.gz
> cd qtkeychain-${QTKEYCHAIN_VERSION}
> cmake -DCMAKE_INSTALL_PREFIX:PATH=/usr/local
> ```

Could be `$PREFIX_DIR` ?

With this, using `sudo` below would become unnecessary. Unless there is a reason to install it system-wide?

---

_[src/loginscreen.ui, line 211 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPQzhNKB4Y2fj5uNXKz:-KPQzhNLAQ0ECRGU-KCM:b-txnz6a) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/loginscreen.ui#L211)):_

> ``` XML
>           <widget class="QCheckBox" name="securelyStorePassword2CB">
>            <property name="text">
>             <string>Securely store password</string>
> ```

This is misleading – should be something like `Store password in keychain` instead. Same goes for other occurrences.

---

_[src/persistence/settings.cpp, line 141 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgVCwAjJ0u4O1nkNwg:-KPgVCwAjJ0u4O1nkNwh:b-a25i5) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/persistence/settings.cpp#L141)):_

> ``` cpp
>         autoLogin = s.value("autoLogin", false).toBool();
> #ifdef QTOX_QTKEYCHAIN
>         securelyStorePassword = s.value("securelyStorePassword", false).toBool();
> ```

Also misleading in code.

---

_[src/widget/passwordstorage.h, line 21 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgWkO9gL175n3XJ2b9:-KPgWkO9gL175n3XJ2bA:b-nurx3x) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/passwordstorage.h#L21)):_

> ``` objc
> 
> 
> #ifdef QTOX_QTKEYCHAIN
> ```

Not needed, since it's included only if `QTOX_QTKEYCHAIN` is defined anyway. Same goes for `.cpp` file.

---

_[src/widget/form/profileform.cpp, line 593 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgZKkj6XMFdgluuzVS:-KPgZKkj6XMFdgluuzVT:b-dw6vck) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/profileform.cpp#L593)):_

> Storing the password failed: 

Should say that adding password to the keychain failed. Same goes for the same text in `loginscreen.cpp`.

---

_[src/widget/form/profileform.cpp, line 601 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgZI7RgXc61-pexKlU:-KPgZI7RgXc61-pexKlV:b-kd2rpp) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/profileform.cpp#L601)):_

> Deleting the password failed:

Should say that deleting password from the keychain failed. Same goes for the same text in `privacyform.cpp`.

---

_[src/widget/form/settings/privacyform.cpp, line 28 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgVqObLOM0yg40qsA2:-KPgVqObLOM0yg40qsA3:b-5o84rg) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/settings/privacyform.cpp#L28)):_

> ``` cpp
> //#include "src/widget/gui.h"
> //#include "src/widget/form/setpassworddialog.h"
> //#include "src/widget/translator.h"
> ```

If commented out stuff is not needed anymore, it should be removed.

---

_[src/widget/form/settings/privacysettings.ui, line 68 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgW3JQulotUVbumesX:-KPgW3JQulotUVbumesY:b-ibyjag) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/settings/privacysettings.ui#L68)):_

> ``` XML
>         <widget class="QPushButton" name="pbClearSecurePasswordStorage">
>          <property name="toolTip">
>           <string>Clear this profile's password from the operating system's or desktop environment's secure password storage</string>
> ```

The `operating system's or desktop environment's` part could be removed with no loss to the info passed, right?

Also, I wonder if changing `secure password storage` to `keychain` would be better?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3332#-:-KPgY6V5MU6BqwQI8OyT:b-7u8icd)_

<!-- Sent from Reviewable.io -->

initramfs commented:
Review status: 2 of 17 files reviewed at latest revision, 13 unresolved discussions.

---

_[src/widget/loginscreen.h, line 71 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgcCQZBxg7jP7rnJ9F:-KPgcCQ_vXeLlZhe_Iq4:bq9hoal) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/loginscreen.h#L71)):_

> ``` objc
>     void checkCapsLock();
> #ifdef QTOX_QTKEYCHAIN
>     void recallPassword(const QString &profileName);
> ```

According to the style guide, this should be `const QString& profileName`. Same goes for all other occurences.

---

_[src/widget/passwordstorage.cpp, line 24 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgcX2MkCy-IiWLxteo:-KPgcX2MkCy-IiWLxtep:b-mti0h1) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/passwordstorage.cpp#L24)):_

> ``` cpp
> #include "passwordstorage.h"
> 
> static const QString serviceName(QStringLiteral("qToX"));
> ```

Is there any reason for the capital 'X' in "qToX"? `qTox` is stylized with a lowercase 'x'.

---

_[src/widget/form/settings/privacyform.cpp, line 99 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgdPf_vXMzCH-C_XOV:-KPgdPfaT12o5bjBOaR2:b-xhovpu) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/settings/privacyform.cpp#L99)):_

> ``` cpp
> void PrivacyForm::deletePassword()
> {
>     QKeychain::Job *job = new DeleteToXPasswordJob(Nexus::getProfile()->getName(), this);
> ```

This memory doesn't seem to ever be deleted. Memory leak?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3332)_

<!-- Sent from Reviewable.io -->

thfi commented:
Review status: 2 of 17 files reviewed at latest revision, 14 unresolved discussions.

---

_[qtox.pro, line 71 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPQzJHWEiGpMVXPCIf-:-KPxhOyfZTdoA20d2ELL:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/qtox.pro#L71)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> can be simplified to
> 
> ```
> !contains(DISABLE_QTKEYCHAIN, YES) {
>     DEFINES += QTOX_QTKEYCHAIN
> }
> ```
> 
> </details>
> Done.

---

_[qtox.pro, line 84 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgWzr0XXok_mzpGK5j:-KPxhbtGreO_-jjl4Fit:b6ikpo8) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/qtox.pro#L84)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> This is going to horribly break whenever brew decides to bump qtkeychain version. Is it possible to not hardcode version?
> </details>
> Currently testing a new approach in Travis CI

---

_[.travis/build-ubuntu-14-04.sh, line 58 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgX_qpxzin5FVd61Fa:-KPxhjeN3x7UA4eHE-rB:b34tuir) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/.travis/build-ubuntu-14-04.sh#L58)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> Could be `$PREFIX_DIR` ?
> 
> With this, using `sudo` below would become unnecessary. Unless there is a reason to install it system-wide?
> </details>
> $PREFIX_DIR is set for ffmpeg. I rewrote both the test for QtKeychain and ffmpeg. Travis will tell me if it breaks or if it works.

---

_[src/loginscreen.ui, line 211 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPQzhNKB4Y2fj5uNXKz:-KPxhzutqYfzjtF_-iAs:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/loginscreen.ui#L211)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> This is misleading – should be something like `Store password in keychain` instead. Same goes for other occurrences.
> </details>
> Done.

---

_[src/persistence/settings.cpp, line 141 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgVCwAjJ0u4O1nkNwg:-KPxi14dY2db9B_VvV1G:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/persistence/settings.cpp#L141)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> Also misleading in code.
> </details>
> Done.

---

_[src/widget/loginscreen.h, line 71 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgcCQZBxg7jP7rnJ9F:-KPxi4LljPL8jzfO6q0k:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/loginscreen.h#L71)):_
<details><summary>

<i>Previously, initramfs wrote…</i></summary>



> According to the style guide, this should be `const QString& profileName`. Same goes for all other occurences.
> </details>
> Done.

---

_[src/widget/passwordstorage.cpp, line 24 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgcX2MkCy-IiWLxteo:-KPxi6PX9md96KA_tXTa:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/passwordstorage.cpp#L24)):_
<details><summary>

<i>Previously, initramfs wrote…</i></summary>



> Is there any reason for the capital 'X' in "qToX"? `qTox` is stylized with a lowercase 'x'.
> </details>
> Done.

---

_[src/widget/passwordstorage.h, line 21 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgWkO9gL175n3XJ2b9:-KPxi7waJTahgC02E-yw:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/passwordstorage.h#L21)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> Not needed, since it's included only if `QTOX_QTKEYCHAIN` is defined anyway. Same goes for `.cpp` file.
> </details>
> Done.

---

_[src/widget/form/profileform.cpp, line 593 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgZKkj6XMFdgluuzVS:-KPxiA39LKwi_venKKnu:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/profileform.cpp#L593)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> > Storing the password failed: 
> 
> Should say that adding password to the keychain failed. Same goes for the same text in `loginscreen.cpp`.
> 
> </details>
> Done.

---

_[src/widget/form/profileform.cpp, line 601 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgZI7RgXc61-pexKlU:-KPxiBEaTFmfnWeLbgxd:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/profileform.cpp#L601)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> > Deleting the password failed:
> 
> Should say that deleting password from the keychain failed. Same goes for the same text in `privacyform.cpp`.
> 
> </details>
> Done.

---

_[src/widget/form/settings/privacyform.cpp, line 28 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgVqObLOM0yg40qsA2:-KPxi_mxrq4tniSfXySF:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/settings/privacyform.cpp#L28)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> If commented out stuff is not needed anymore, it should be removed.
> </details>
> Done.

---

_[src/widget/form/settings/privacyform.cpp, line 97 [r2]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPxibNizQHzkVrCsD-w:-KPxibNizQHzkVrCsD-x:b9pu6zl) ([raw file](https://github.com/qtox/qtox/blob/ecd5e0a7579c56e9771bc98e6859a72d1723f325/src/widget/form/settings/privacyform.cpp#L97)):_

> ``` cpp
>     QKeychain::Job *job = new DeleteToXPasswordJob(Nexus::getProfile()->getName(), this);
>     connect(job, &DeleteToXPasswordJob::finished, this, &PrivacyForm::onDeletePasswordFinished);
>     job->start();
> ```

Actually no. Those various *Job classes inherit from QKeychain::Job which calls on itself deleteLater() once it is finished.

---

_[src/widget/form/settings/privacysettings.ui, line 68 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgW3JQulotUVbumesX:-KPxirrX3s__70X36wJ_:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/settings/privacysettings.ui#L68)):_
<details><summary>

<i>Previously, zetok wrote…</i></summary>



> The `operating system's or desktop environment's` part could be removed with no loss to the info passed, right?
> 
> Also, I wonder if changing `secure password storage` to `keychain` would be better?
> </details>
> Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3332)_

<!-- Sent from Reviewable.io -->

zetok commented:
Review status: 2 of 17 files reviewed at latest revision, 6 unresolved discussions.

---

_[src/widget/loginscreen.cpp, line 355 [r2]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPxzCNuISgmict4axTb:-KPxzCNuISgmict4axTc:bjzhi43) ([raw file](https://github.com/qtox/qtox/blob/ecd5e0a7579c56e9771bc98e6859a72d1723f325/src/widget/loginscreen.cpp#L355)):_

> ``` cpp
>     WriteToXPasswordJob *job = static_cast<WriteToXPasswordJob*>(sender());
>     if (job->error())
>         GUI::showWarning(tr("Storing password in keychain failed", "Title of QtKeychain error message"), tr("Storing the password in keychain failed: %1", "text of QtKeychain error message").arg(job->errorString()));
> ```

This line could be split into at least 2. Generally try to be within 80 chars per line. 100-120 chars is still acceptable, just not as good. Longer stuff should be just split.

---

_[src/widget/form/profileform.cpp, line 601 [r1]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPgZI7RgXc61-pexKlU:-KPy-KANjGKQFKzNnX_w:b1d2iiv) ([raw file](https://github.com/qtox/qtox/blob/5ca0b7caf4fd7ac49a29c680e35ef8fb1251ab6e/src/widget/form/profileform.cpp#L601)):_

> Deleting the password in keychain failed: 

Um, I don't English natively, but wouldn't `from keychain` sound more natural than `in keychain` ?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3332)_

<!-- Sent from Reviewable.io -->

initramfs commented:
Review status: 2 of 17 files reviewed at latest revision, 4 unresolved discussions.

---

_[src/widget/form/settings/privacyform.cpp, line 97 [r2]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KPxibNizQHzkVrCsD-w:-KQH2983xX95ypZLgaWc:b670y77) ([raw file](https://github.com/qtox/qtox/blob/ecd5e0a7579c56e9771bc98e6859a72d1723f325/src/widget/form/settings/privacyform.cpp#L97)):_
<details><summary>

<i>Previously, thfi wrote…</i></summary>



> Actually no. Those various *Job classes inherit from QKeychain::Job which calls on itself deleteLater() once it is finished.
> </details>
> OK. Didn't know that.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3332)_

<!-- Sent from Reviewable.io -->

sudden6 commented:
Reviewed 1 of 17 files at r1, 3 of 13 files at r2, 4 of 9 files at r3.
Review status: 10 of 17 files reviewed at latest revision, 9 unresolved discussions.

---

_[src/widget/loginscreen.cpp, line 352 [r3]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KQbNJoUepJOyvMck6Ys:-KQbNJoUepJOyvMck6Yt:bbo4dt2) ([raw file](https://github.com/qtox/qtox/blob/3069f7f1199916600b70107a9a83eb8c8486c07b/src/widget/loginscreen.cpp#L352)):_

> ``` cpp
> }
> 
> void LoginScreen::onStorePasswordFinished(){
> ```

Coding style, the opening `{` bracket should be on a new line

---

_[src/widget/passwordstorage.cpp, line 25 [r3]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KQbNZKfNGLKvnjSm_QR:-KQbNZKfNGLKvnjSm_QS:b-jxvh92) ([raw file](https://github.com/qtox/qtox/blob/3069f7f1199916600b70107a9a83eb8c8486c07b/src/widget/passwordstorage.cpp#L25)):_

> ``` cpp
> 
> ReadToXPasswordJob::ReadToXPasswordJob(const QString& profileName, QObject *parent)
>     : QKeychain::ReadPasswordJob(serviceName, parent){
> ```

`{` coding style in the whole file

---

_[src/widget/passwordstorage.h, line 26 [r3]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KQbNnQtCsCQd3JeCb6p:-KQbNnQuYLE_3V7Sw1GS:b-dj8bad) ([raw file](https://github.com/qtox/qtox/blob/3069f7f1199916600b70107a9a83eb8c8486c07b/src/widget/passwordstorage.h#L26)):_

> ``` objc
> #include <qt5keychain/keychain.h>
> 
> class ReadToXPasswordJob : public QKeychain::ReadPasswordJob {
> ```

`{` coding style in whole file

---

_[src/widget/form/profileform.cpp, line 588 [r3]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KQbOKu60JB49-uANPhR:-KQbOKu60JB49-uANPhS:bnxhurk) ([raw file](https://github.com/qtox/qtox/blob/3069f7f1199916600b70107a9a83eb8c8486c07b/src/widget/form/profileform.cpp#L588)):_

> ``` cpp
> }
> 
> void ProfileForm::onStorePasswordFinished(){
> ```

coding style

---

_[src/widget/form/profileform.cpp, line 598 [r3]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KQbOP_o8zjzhhbn5Q3V:-KQbOP_o8zjzhhbn5Q3W:bnxhurk) ([raw file](https://github.com/qtox/qtox/blob/3069f7f1199916600b70107a9a83eb8c8486c07b/src/widget/form/profileform.cpp#L598)):_

> ``` cpp
> }
> 
> void ProfileForm::onDeletePasswordFinished(){
> ```

coding style

---

_[src/widget/form/settings/privacyform.cpp, line 53 [r3]](https://reviewable.io:443/reviews/qtox/qtox/3332#-KQbPmBnOB2aeePPNtjX:-KQbPmBojYX4Oedrmb13:b-j35zh) ([raw file](https://github.com/qtox/qtox/blob/3069f7f1199916600b70107a9a83eb8c8486c07b/src/widget/form/settings/privacyform.cpp#L53)):_

> ``` cpp
>     eventsInit();
> 
> // TODO refactor into eventsInit()?
> ```

No need to refactor into `eventsInit()`, rather use http://doc.qt.io/qt-4.8/designer-using-a-ui-file.html#widgets-and-dialogs-with-auto-connect like in all other settings forms.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3332)_

<!-- Sent from Reviewable.io -->

sudden6 commented:
Reviewed 1 of 13 files at r2, 5 of 7 files at r4.
Review status: 14 of 17 files reviewed at latest revision, 3 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3332)_

<!-- Sent from Reviewable.io -->

zetok commented:
@thfi btw, it would be good to at least comment that you've fixed/changed stuff, so that people would be notified that things need to be looked at again.

If you don't mind using reviewable, it does provide a nice `Done` button, it nicely marks stuff that was changed.

sudden6 commented:
The last thing that is needed for that PR would be documentation in https://github.com/qTox/qTox/blob/master/doc/user_manual_en.md

But IMHO you could do that in a follow up PR.

Also please comment to notify, I overlooked your changes because of that...

---

Review status: 14 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3332#-:-KRjLhUmm6NKLp7VDdSs:bym6rmh)_

<!-- Sent from Reviewable.io -->

zetok commented:
on f2fc3737c0a0a1aef9ef4f8a00e16cfce349f1a6:
- [ ] depending on DPI, text on login screen can be cut, it could be a lot worse with translations:
  ![spectacle jg9493](https://cloud.githubusercontent.com/assets/3148759/18565070/3fb07d08-7b86-11e6-8e6a-c6dae76c0114.png)
  
  DPI: 104
- [ ] when I've tried to store a password in keychain for a profile that doesn't have a password:
  ![spectacle fv9493](https://cloud.githubusercontent.com/assets/3148759/18565243/e9f0d3b2-7b86-11e6-8e33-353b3e63075e.png)
  
  profile name (perhaps relevant): `.tox` (whitespace dot tox)
- [ ] when trying to "remove" "password" in settings from the above mentioned profile:
  ![spectacle ig9493](https://cloud.githubusercontent.com/assets/3148759/18565308/37e24042-7b87-11e6-9f08-bd06ce682511.png)
- [ ] since it doesn't seem to work at all (even with profiles that are actually encrypted) there are probably missing docs

thfi commented:
> @thfi btw, it would be good to at least comment that you've fixed/changed stuff, so that people would be notified that things need to be looked at again.
> If you don't mind using reviewable, it does provide a nice Done button, it nicely marks stuff that was changed.

Sorry, this is the first time I am using Reviewable. I think I pressed Done at some occasions, but missed others.

thfi commented:
> when I've tried to store a password in keychain for a profile that doesn't have a password:
> when trying to "remove" "password" in settings from the above mentioned profile:
> "Failed to execute program org.kde.kwalletd"

@zetok  The error message you see is in part coming from QtKeychain. It looks like QtKeychain has problems talking to your KWallet installation. If you are running the most recent stable version of QtKeychain and KWallet is working in other applications, then you should report that issue upstream to QtKeychain. The API provided by QtKeychain is backend-agnostic, so it is not possible to trace back this problem in detail from my patch's end of the line.

> depending on DPI, text on login screen can be cut, it could be a lot worse with translations

I changed the text to "Store password" and set the check box's tool tip to "Store password in keychain" for both instances. This change will be part of the revised pull request as of commit 410589d.

zetok commented:
> It looks like QtKeychain has problems talking to your KWallet installation.

Nah, I don't even have `kde-apps/kwalletd` installed.

> If you are running the most recent stable version of QtKeychain and KWallet is working in other applications, then you should report that issue upstream to QtKeychain.

I have lowest supported QtKeychain, i.e. `0.6.2`.

IMO if something has no chances of working in the first place, qTox's interface shouldn't show an option for it.
QtKeychain has an API that could say `"Yeah, no, this ain't gonna fly.."`, right? If so, it should be used to determine whether options should be enabled in the first place.

zetok commented:
@thfi any update on the status of the PR?

zetok commented:
PR didn't have any activity for over 2 months, and no one seems to be interested in working on it, thus closing.

If that's not the case, please either ask to reopen the pull, or make a new one.

za3k commented:
Not sure I agree that keychains should be off by default as a blanket policy, I think a per-keychain-system default makes more sense. For example on OS X, they should probably be on by default, because they're easy to use without instruction (just enter the system password) and are used in many problems. On KDE Wallet, off by default probably makes sense, because lots of people with KDE Wallet installed don't even use it. And I don't use Windows enough to comment there.
